OGC Api features suport for non 4326 layers

domi4484 opened the pull request:
Read layer CRS from collections request instead of hard coded

fixes #50051

domi4484 committed:
Wfs3 read layer CRS from collections request

domi4484 committed:
Dont check with a hard coded crs assumption

domi4484 committed:
Keep bounding check for geographic crs

domi4484 committed:
Update src/providers/wfs/oapif/qgsoapifprovider.cpp

Co-authored-by: Denis Rouzaud <denis.rouzaud@gmail.com>

domi4484 committed:
If CRS is not announced by the server use CRS84 as specified by OGC

domi4484 committed:
Add some tests

domi4484 committed:
Make linter happy

rouault commented on src/providers/wfs/oapif/qgsoapifprovider.cpp:
```diff
@@ -702,27 +714,32 @@ void QgsOapifFeatureDownloaderImpl::run( bool serializeFeatures, long long maxFe
-                   qgsDoubleToString( miny ),
-                   qgsDoubleToString( maxx ),
-                   qgsDoubleToString( maxy ) );
+             .arg( qgsDoubleToString( rect.xMinimum() ),
```
if the CRS of the layer is not the default OGC:CRS84 one, then the bbox-crs parameter should be specified (cf https://docs.opengeospatial.org/is/18-058r1/18-058r1.html#_parameter_bbox_crs), and care should be taken to do the proper axis order inversion for CRS with latitude-longitude or northing-easting order. Cf requirement 23.B of https://docs.opengeospatial.org/is/17-069r4/17-069r4.html#_parameter_bbox that mentions in a rather cryptic way that the first value should be "Lower left corner, coordinate axis 1". Slide 9 of https://files.justobjects.nl/presentation/geonovum-2021-api-strategie/OGC-API-Features.pdf makes it more explicit: "bbox-crs: URI of the CRS in which the coordinates are in the bbox parameter; axis order follows the axis order in the CRS"

domi4484 replied:
Is there a way to get the CRS in the format with the url? Or could I add a function in `QgsCoordinateReferenceSystem` to return a string in the format like `http://www.opengis.net/def/crs/OGC/1.3/CRS84`? A problem is that the version is not parsed...

About the axis order, from what I could see the currentRect is in the layer CRS, respecting also axis order inverted or not. In that case the clamping would be wrong for crs with inverted axis. I'll check that

elpaso replied:
@domi4484 have a look to QString QgsServerApiUtils::crsToOgcUri

elpaso replied:
If that suits your needs you might consider moving it from server to core.

domi4484 replied:
Thanks @elpaso , I moved it to QgsCoordinateReferenceSystem and declared the one in QgsServerApiUtils deprecated.

@rouault bbox-crs paramter in request is implemented in the last commit

rouault replied:
> bbox-crs paramter in request is implemented in the last commit

axis inversion should also be added if QgsCoordinateReferenceSystem::hasAxisInverted()  ( to be checked how it behaves with OGC:CRS84, where axis order is longitude, latitude, contrary to EPSG:4326 which is latitude, longitude)

domi4484 replied:
Ok now it should be good. At the beginning I thought erroneously that the received rectangle was already right inverted.

domi4484 commented on src/providers/wfs/oapif/qgsoapifcollection.cpp:
```diff
@@ -297,6 +298,19 @@ bool QgsOapifCollection::deserialize( const json &j )
+      for ( const auto &crsUrl : crsUrls )
+      {
+        QString crs = QString::fromStdString( crsUrl.get<std::string>() );
+        mLayerMetadata.setCrs( QgsCoordinateReferenceSystem::fromOgcWmsCrs( crs ) );
```
No, in fact I think the first would make more sense, until support for multiple is implemented.

domi4484 committed:
Apply suggestions from code review

Co-authored-by: Even Rouault <even.rouault@spatialys.com>

domi4484 committed:
Use first CRS of the list

domi4484 committed:
Add bbox-crs parameter to the request (if not default CRS84)

rouault commented on src/core/proj/qgscoordinatereferencesystem.cpp:
```diff
@@ -1495,6 +1495,29 @@ QgsRectangle QgsCoordinateReferenceSystem::bounds() const
+  if ( parts.length() == 2 )
+  {
+    if ( parts[0] == QLatin1String( "EPSG" ) )
+      return  QStringLiteral( "http://www.opengis.net/def/crs/EPSG/9.6.2/%1" ).arg( parts[1] ) ;
```
the usual practice is to use the dummy 0 generic version:
```suggestion
      return  QStringLiteral( "http://www.opengis.net/def/crs/EPSG/0/%1" ).arg( parts[1] ) ;
```

http://www.opengis.net/def/crs/EPSG/0/4326 is a dereferencable URL, whereas http://www.opengis.net/def/crs/EPSG/9.6.2/4326 is not

domi4484 replied:
Ok thanks, and I suppose for the OGC url the same?

rouault replied:
> and I suppose for the OGC url the same?

hum, I'm more used in seeing 1.3 being used... But both http://www.opengis.net/def/crs/OGC/0/CRS84 and http://www.opengis.net/def/crs/OGC/1.3/CRS84 work

rouault commented on tests/src/python/test_provider_oapif.py:
```diff
@@ -326,7 +334,7 @@ def testBbox(self):
             ]
         }
-        with open(sanitize(endpoint, '/collections/mycollection/items?limit=1000&' + ACCEPT_ITEMS), 'wb') as f:
+        with open(sanitize(endpoint, '/collections/mycollection/items?limit=1000&bbox=-180,-90,180,90&bbox-crs=http://www.opengis.net/def/crs/EPSG/9.6.2/4326&' + ACCEPT_ITEMS), 'wb') as f:
```
for EPSG:4326, that should be bbox=-90,-180,90,180

domi4484 committed:
Update src/core/proj/qgscoordinatereferencesystem.cpp

Co-authored-by: Even Rouault <even.rouault@spatialys.com>

domi4484 committed:
Invert bbox rect if crs has inverted axis

domi4484 committed:
Fix test

domi4484 committed:
Fix tests

domi4484 commented:
@rouault I addressed all the points. could you please redo a review?


> the usual practice is to use the dummy 0 generic version:
> 
> http://www.opengis.net/def/crs/EPSG/0/4326 is a dereferencable URL, whereas http://www.opengis.net/def/crs/EPSG/9.6.2/4326 is not

@elpaso I changed everywhere the version 9.6.2 of the url to 0.

elpaso commented:
> @rouault I addressed all the points. could you please redo a review?
> 
> > the usual practice is to use the dummy 0 generic version:
> > http://www.opengis.net/def/crs/EPSG/0/4326 is a dereferencable URL, whereas http://www.opengis.net/def/crs/EPSG/9.6.2/4326 is not
> 
> @elpaso I changed everywhere the version 9.6.2 of the url to 0.

Sounds good to me, it shouldn't break anything.

rouault commented:
> Sorry to ask question after the PR was merged but does QGIS check the conformance class before trying to use crs in API requests?

no it doesn't. But it should normally only use the extension if the server advertizes a non CRS84 crs in the "crs" array member. What does your /collection response look like ?

vincentsarago commented:
oh interesting, I guess it's because we return EPSG:4326 instead of CRS84 https://github.com/developmentseed/tipg/blob/main/tipg/model.py#L89
